File: Data Structures & Algorithms/two-integer-sum/submission-0.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        vals = {}
        for i in range(len(nums)):
            if nums[i] not in vals.keys():
                vals[target - nums[i]] = i
            else:
                return [vals.get(nums[i]), i]

File: Data Structures & Algorithms/combination-sum-iv/submission-0.java
class Solution {
    private Map<Integer, Integer> memo;
    public int combinationSum4(int[] nums, int target) {
        Arrays.sort(nums);
        memo = new HashMap<>();
        memo.put(0, 1);
        return dfs(nums, target);
    }

    private int dfs(int[] nums, int remaining){
        if(memo.containsKey(remaining)){
            return memo.get(remaining);
        }
        int res = 0;
        for (int num: nums){
            if(remaining < num) {
                break;
            }
            res += dfs(nums, remaining - num);
        }
        memo.put(remaining, res);
        return res;
    }
}